DIPs/DIP1009.md

```diff
@@ -42,38 +44,51 @@ body
+2. Contracts require the use of the keyword `do` (or `body`, see [DIP1003](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1003.md)) before the function body. Any way of writing them that eliminated this need would be an improvement.
+
+3. `in` and `out` contracts in D require explicitly writing `assert` statements, or some equivalent, in their code block in order to perform checks. But because of the nature of contracts, `assert` can be assumed in the vast majority of cases, and it is tedious to have to write it specifically. For the common case, the mere presence of `in` or `out` is sufficient. For those programmers who require the use of other forms of checking, the existing syntax will remain unchanged. 
 
-History shows that a small change in syntax can dramatically change the usability of a language feature. By adding a new syntax for contracts, an increase in their usage can be expected.
+Experience demonstrates, as in the case of `=>` lambdas, that a small change in syntax can dramatically affect the usage of a language feature. By adding a new syntax for contracts, a similar increase in their usage can be expected.
 
 Furthermore, because the existing `invariant` grammar is so similar to `in` and `out` contract grammar, any improvements to `in` and `out` contracts can be transferred easily to invariants as well. Consistency across language constructs is desirable.
 
 ## Description
 
-Allow `in`, `out`, and `invariant` contracts to be written as parenthesized expressions. Allow multiple expressions of this kind. Create a new grammatical construct called ContractParameters, which copies and replaces the parameters to `assert` expressions. Use the new construct for `in`, `out`, `assert`, and `invariant` expressions alike. `in` and `out` expressions must come at the end of the function declarator suffix, and before the regular contracts, if any. `invariant` expressions require a semicolon terminator. (Body-less interface functions which use only the new syntax also require a semicolon terminator.) The keyword `do` or `body` is not required after the new `in` and `out` expressions.  The choice of syntax for `out` expressions requires a separate analysis (see the Analysis section below). For the sake of the following example, the `out` expression syntax requires a semicolon separating an optional return identifier from its subsequent ContractParameters:
+Take the grammar for existing `assert` parameters, and call it ContractParameters. Copy this grammar and use it in all the new syntaxes.
+
+Introduce two new `in` and `out` contract syntaxes, called InExpression and OutExpression.
+
+An InExpression is `in`, followed by parentheses containing ContractParameters, exactly mimicking an `assert` call.
+
+The recommended OutExpression syntax is `out`, followed by parentheses containing an optional return identifier, a manditory semicolon, and ContractParameters (see Analysis below).
 
```

Comment:
The ContractParameters bit is still not as clear as it could be. Perhaps it's better not to lead off with that. Maybe even avoid it until the grammar section. I'm thinking something like this:

"Modify the language grammar to introduce new, optional syntaxes for `in` and `out` contracts, with the goal of making them more concise and readable by eliminating the need for braces and `assert` statements. The new syntaxes can be referred to by the new grammar symbols InExpression and OutExpression, respectively.

An InExpression is `in`, followed by parentheses containing a Boolean expression. This syntax exactly mimics that of a call to `assert`.

An OutExpression is `out`, followed by parentheses containing, in the listed order: an optional return identifier, a mandatory semicolon (even when the return identifier is absent), and a Boolean expression as found in a call to assert.

Multiple `in` and `out` contracts are allowed on a single function, with one contract per line before the function body. No semicolon is required to terminate a contract, and the keyword indicating the beginning of the function body is optional; the body's opening brace is sufficient."